CASSSIDECAR-360: Sidecar API Endpoint for Nodetool Compaction Stop

shalnisundram opened the pull request:
Add bespoke API endpoint for nodetool compaction stop.

This implementation carries out functionality currently performed when execution command `nodetool stop`, which is a short-running operation that requests ongoing compactions of a specified` type` or `id` to be stopped.
 
See [CASSANDRA-20021 JIRA ticket](https://issues.apache.org/jira/browse/CASSSIDECAR-360) - more details can be found on ongoing effort to build APIs for nodetool C* cluster management operations.

patch by Shalni Sundram

Commit:
implement API endpoint for nodetool compaction stop

Commit:
include CassandraCompactionManagerOperations.java modification for nodetool compaction stop API implementation

bbotella commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +53,27 @@ public List<Map<String, String>> getCompactions()
+        else
+        {
+            // Handler guarantees compactionType is non-null/non-empty
+            proxy.stopCompaction(compactionType);
```
Maybe we should add a @NotNull annotation to compactionType?

shalnisundram replied:
See change on [CassandraCompactionManagerOperations line 77 ](https://github.com/shalnisundram/cassandra-sidecar/blob/6d42f45dbdbbc30da2149990272d2494ca437227/adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java#L77). Not exactly an annotation but a requireNonNull check. Let me know if that's acceptable.

bbotella commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/ApiEndpointsV1.java:
```diff
@@ -143,7 +143,7 @@ public final class ApiEndpointsV1
     public static final String STREAM_STATS_ROUTE = API_V1 + CASSANDRA + "/stats/streams";
     public static final String TABLE_STATS_ROUTE = API_V1 + CASSANDRA + PER_KEYSPACE + PER_TABLE + "/stats";
-
+    public static final String COMPACTION_STOP_ROUTE = API_V1 + CASSANDRA + "/operations/compaction/stop";
```
nit: I see that `/operations/` is repeated across some routes. What about extracting it to a static final?

bbotella commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/request/CompactionStopRequest.java:
```diff
@@ -0,0 +1,46 @@
+import org.apache.cassandra.sidecar.common.response.OperationalJobResponse;
+
+/**
+ * Represents a request to execute node decommission operation
```
Is node decommission the same as compaction stop?

shalnisundram replied:
they're not the same that's a typo - thanks for catching. Fixed.

bbotella commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/request/data/CompactionStopRequestPayload.java:
```diff
@@ -0,0 +1,82 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CompactionStopRequestPayload
+{
+    private final String compactionType;
```
Should this be an enum instead of a free string?

shalnisundram replied:
Chose deliberately to use `String`, as I figured this would encourage forward compatibility with future C* versions if more compactionTypes are added later.  i.e., would want to avoid sidecar forcing all C* versions into a single enum definition, and would rather leave this responsibility to cassandra itself. Would appreciate @arjunashok's input for this though.

sklaha replied:
You are already defining the compaction types as a set in AbstractHandler.SUPPORTED_COMPACTION_TYPES. Better to make it an enum instead of a set.

shalnisundram replied:
Done

bbotella commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStopResponse.java:
```diff
@@ -0,0 +1,225 @@
+    public static final String COMPACTION_ID = "compaction_id";
+    public static final String STATUS = "status";
+    public static final String ERROR_CODE = "error_code";
+    public static final String REASON = "reason";
```
nit: What about appending `_KEY` to the constant names?

bbotella commented on server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java:
```diff
@@ -21,37 +21,14 @@
-import org.apache.cassandra.sidecar.handlers.StreamStatsHandler;
-import org.apache.cassandra.sidecar.handlers.TableStatsHandler;
-import org.apache.cassandra.sidecar.handlers.TokenRangeReplicaMapHandler;
+import org.apache.cassandra.sidecar.handlers.*;
```
Please revert this. We try to avoid the * imports

shalnisundram replied:
Thanks for letting me know - fixed.

bbotella commented on server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java:
```diff
@@ -109,6 +86,25 @@ VertxRoute cassandraCompactionStatsRoute(RouteBuilder.Factory factory,
+    @APIResponse(description = "Compaction stop operation completed",
+            responseCode = "200",
+            content = @Content(mediaType = "application/json",
+                    schema = @Schema(implementation = CompactionStopResponse.class)))
```
Please add the @APIResponse for the error codes as well

shalnisundram replied:
See [lines 149-189 of CompactionStopHandler.java](https://github.com/shalnisundram/cassandra-sidecar/blob/6d42f45dbdbbc30da2149990272d2494ca437227/server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java#L149) - I do some custom error handling here, as particular api response codes will have different messages thrown depending on the input. Let me know if I should still edit this section in CassandraOperationsModule.

Commit:
clean imports, add annotations, improve naming for constants

sklaha reviewed (commented):
Thank you for the PR. Looks good at a high level. Please run `./gradlew clean; ./gradlew build` which will show you check style issues. I am leaving my other comments inline.

sklaha commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,190 @@
+                .status("SUCCESS")
+                .errorCode("200 OK")
+                .reason("Operation Succeeded")
+                .build();
```
Do we need these fields? Doesn't HTTP 200 response provide the same information? Also, these fields being set for failure scenarios.

shalnisundram replied:
Some small changes in the handler - the 200 response shouldn't appear upon a failure - these are caught in `extractParamsOrThrow()` on line 133 of this file.
Other than that, we'd decided in the original design to include these fields - they won't be exposed automatically, but let me know if there's a better way of including them in the response.

sklaha commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,190 @@
+     * @return validated CompactionStopRequestPayload
+     */
+    @Override
+    protected CompactionStopRequestPayload extractParamsOrThrow(RoutingContext context)
```
What happens when the request has both compaction id and compaction type?

sklaha replied:
Looks like compaction id takes precedence. It could be better to just fail the request with appropriate error message.

shalnisundram replied:
Done - split into two functions

sklaha commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,27 @@ public List<Map<String, String>> getCompactions()
+     * {@inheritDoc}
+     */
+    @Override
+    public void stopCompaction(String compactionId, String compactionType)
```
I would prefer to have two separate methods for compaction operations based on ID and type, rather than having a single method that accepts two parameters and actually uses only one.

shalnisundram replied:
See comment to previous message above - functions split with `stopById()` taking precedence.

sklaha commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStopIntegrationTest.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    @Test
+    void testStopCompactionByIdSuccess()
```
In the current implementation, a request with a non-existent compaction ID silently succeeds. The JMX call simply succeeds without any action being taken. It would be more informative to update the Cassandra JMX implementation to return a boolean value or some other indicator. I will seek input from others on this.

https://github.pie.apple.com/slaha/aci-cassandra/blob/38ab7d02e0979074f3de1c7f90ce70668d2492b9/src/java/org/apache/cassandra/db/compaction/CompactionManager.java#L2306

sklaha commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStopIntegrationTest.java:
```diff
@@ -0,0 +1,218 @@
+/**
+ * Integration tests for the Compaction Stop API endpoint
+ */
+class CompactionStopIntegrationTest extends SharedClusterSidecarIntegrationTestBase
```
Can we add a test that triggers compaction and effectively stops ongoing compactions using the API? You can get an idea of how to trigger a compaction from the `testCompactionStatsRetrieval`.

https://github.com/apache/cassandra-sidecar/blob/c987f2956b22b457cf3ac68408657e706d0b9a1b/integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CassandraStatsIntegrationTest.java#L285

shalnisundram replied:
Good idea and thanks for the reference. Added two integration tests in this file - one for stopping compactions by type and the other by ID. Tests check this by comparing list of active compactions before and after new API call.

Made SSTables large + only considered compactions under 50% complete to catch and stop them before they finish on their own.

Commit:
add compaction type enum support, add full compaction integration tests, fix bug in handler

arjunashok commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/ApiEndpointsV1.java:
```diff
@@ -139,16 +140,18 @@ public final class ApiEndpointsV1
-
+    public static final String COMPACTION_STOP_ROUTE = API_V1 + CASSANDRA + OPERATIONS + "/compaction/stop";
+    public static final String NODE_DRAIN_ROUTE = API_V1 + CASSANDRA + OPERATIONS + "/drain";
+    public static final String LIFECYCLE_ROUTE = API_V1 + CASSANDRA + "/lifecycle";
```
These changes are unrelated to your PR, possibly as a result of the merge from the main branch.  You might want to rebase your changes on top of the main/develop branch contents, so the PR only has your changes.

shalnisundram replied:
Rebased - changed file now only includes compaction stop additions related to new compaction stop imports and routes

arjunashok commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/request/CompactionStopRequest.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Represents a request to execute node stop operation
+ */
+public class CompactionStopRequest extends JsonRequest<OperationalJobResponse>
```
Should this be `extends JsonRequest<CompactionStopResponse>`

shalnisundram replied:
Yes - changed

arjunashok commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStopResponse.java:
```diff
@@ -0,0 +1,225 @@
+    public static final String ERROR_CODE_KEY = "error_code";
+    public static final String REASON_KEY = "reason";
+
+    private final String compactionType;
```
This should use the enum for consistency.

shalnisundram replied:
Done

arjunashok commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStopResponse.java:
```diff
@@ -0,0 +1,225 @@
+    public static final String COMPACTION_TYPE_KEY = "compaction_type";
+    public static final String COMPACTION_ID_KEY = "compaction_id";
+    public static final String STATUS_KEY = "status";
+    public static final String ERROR_CODE_KEY = "error_code";
```
Let's evaluate the intended usage of the `error_code` field in determining the values. Based on the current name, it seems to suggest this is a  status code (currently returning `200 OK`) 

The 200 status code would be redundant in this case, as it is implied with the HTTP response code. 

The other option is to do away with this field since we have a `reason` field to denote an arbitrary string reason when there is an error (or a non-200 response).

arjunashok replied:
Also, can `status` be an enum here? Seems like it can be one of 3 or 4 values - `SUBMITTED`, `PENDING` or `FAILED`

shalnisundram replied:
Original status options for the command were just two - "PENDING" and "FAILED". For this API it's "SUBMITTED" (more clear I think) and "FAILED".
Status changed to enum - see CompactionStopStatus.java for new enum class

arjunashok commented on server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java:
```diff
@@ -112,6 +98,27 @@ VertxRoute cassandraCompactionStatsRoute(RouteBuilder.Factory factory,
+            content = @Content(mediaType = "application/json",
+                    schema = @Schema(implementation = CompactionStopResponse.class)))
+    @APIResponse(responseCode = "500", description = "Internal server error")
+    @APIResponse(responseCode = "400", description = "Internal server error on getting the ring view")
```
Is this the appropriate description? `400` implies bad request, so it should not state "Internal server error"

shalnisundram replied:
Yes thanks for catching that. Changed to "Invalid request - malformed JSON body or invalid compaction parameters"

Commit:
remove redundant JSON response fields for compaction stop endpoint

arjunashok commented on client/src/main/java/org/apache/cassandra/sidecar/client/SidecarClient.java:
```diff
@@ -752,6 +753,20 @@ public CompletableFuture<CompactionStatsResponse> compactionStats(SidecarInstanc
     }
 
+    /**
+     * Executes the compaction stats request using the default retry policy and provided {@code instance}.
```
Nit: update to `compaction stop`

arjunashok commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStopIntegrationTest.java:
```diff
@@ -0,0 +1,461 @@
+        assertThat(stopResponse).isNotNull();
+        assertThat(stopResponse.status()).isEqualTo(CompactionStopStatus.SUBMITTED);
+        assertThat(stopResponse.compactionType()).isEqualTo("COMPACTION");
+        assertThat(stopResponse.reason()).isEqualTo("Operation Succeeded");
```
No longer applicable, correct?

shalnisundram replied:
Omitted the reason() field - the status , type, and id fields still stand in the response so left those in

arjunashok commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStopIntegrationTest.java:
```diff
@@ -0,0 +1,461 @@
+                    instance.nodetool("compact", TEST_KEYSPACE, TEST_TABLE.table())
+            );
+        });
+        compactionThread.start();
```
`nodetool compact` is non-blocking by default and will return immediately. We like;y do not need to run this on a separate thread.

shalnisundram replied:
gotcha - omitted separate thread.

arjunashok commented on integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStopIntegrationTest.java:
```diff
@@ -0,0 +1,461 @@
+                    // ~1GB+ in a few seconds)
+                    boolean compactionStopped = false;
+                    for (int verifyAttempt = 0; verifyAttempt < 10; verifyAttempt++) {
+                        Thread.sleep(300);
```
This makes the test brittle and can introduce flakiness. Lets use `AssertionUtils.loopAssert()` instead for both the polling loop and for the verification.

shalnisundram replied:
Tried with `loopAssert`, but the tests will typically fail since a condition must be met to pass. Since trying to catch an active compaction will usually not succeed, I kept the manual loop structure in. 
To improve the test, added an `assumeTrue()` at the end of the test so it doesn't pass by default when not finding an active compaction. Now the test is ignored after all attempts: https://github.com/apache/cassandra-sidecar/blob/def6e6fb36cb7fa05bf9e3a0237d8a407cd3e8cc/integration-tests/src/integrationTest/org/apache/cassandra/sidecar/routes/CompactionStopIntegrationTest.java#L451

arjunashok reviewed (commented):
Looking good for the most part. Added some minor comments.

Commit:
add SidecarClientTest test for compaction stop, assumeTrue() for integration test

bbotella commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,32 @@ public List<Map<String, String>> getCompactions()
+    @Override
+    public void stopCompactionById(String compactionId)
+    {
+        CompactionManagerJmxOperations proxy = jmxClient.proxy(
```
Do you need to declare the proxy if compactionId == null? Can't we do this inside the if?

shalnisundram replied:
Good catch yes - changed.

bbotella commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/ApiEndpointsV1.java:
```diff
@@ -143,7 +143,7 @@ public final class ApiEndpointsV1
     public static final String STREAM_STATS_ROUTE = API_V1 + CASSANDRA + "/stats/streams";
     public static final String TABLE_STATS_ROUTE = API_V1 + CASSANDRA + PER_KEYSPACE + PER_TABLE + "/stats";
-
+    public static final String COMPACTION_STOP_ROUTE = API_V1 + CASSANDRA + "/operations/compaction/stop";
```
nit: I see the `operations` part of the path in several static variables here. Let's extract it to its own private static

shalnisundram replied:
Done. Added the private static, but only used it for the `node stop` route rather than `node drain` and `node decommission` routes which also uses this path, as those changes would not be directly related to this PR.

bbotella commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/request/data/CompactionStopRequestPayload.java:
```diff
@@ -0,0 +1,82 @@
+        return "CompactionStopRequestPayload{" +
+               "compactionType='" + compactionType + '\'' +
+               ", compactionId='" + compactionId + '\'' +
+               '}';
```
```suggestion
        return "CompactionStopRequestPayload{" +
               "compactionType='" + compactionType + "'"  +
               ", compactionId='" + compactionId + "'"  +
               "}";
```

bbotella commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,169 @@
+
+        // Attempt to stop the compaction
+        // If compactionId is provided, use it (takes precedence over type)
+        if (compactionId != null && !compactionId.trim().isEmpty()) {
```
I see this pattern of "if compactionID then do otherwise do by type" in several different places. What do you think about abstracting that inside operations, with a generic method that does `stopCompaction(@Nullable String compactionId, @Nullable String compactionType)`, and performs that if internally?

Sorry if this has already been discussed.

shalnisundram replied:
Yeah in the latter case we would have a single method that accepts two params but only uses one, which others expressed was not preferable. Thus, preserved the `stopByID()` and `stopCompaction()` structure already used by `node stop`

bbotella commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,169 @@
+
+        // Validate that at least one field is provided
+        if (payload.compactionType() == null &&
+            (payload.compactionId() == null || payload.compactionId().trim().isEmpty()))
```
I also see this pattern of `== null` or `== null or empty` across different places. What about adding boolean public methods on payload that return if they are valid? Something like `isCompactionTypeValid()` and `isCompactionIdValid`?

shalnisundram replied:
Done - added those functions in `CompactionStopRequestPayload.java`, although didn't use those functions for the pattern present in `CassandraCompactionManagerOperations.java`, as those methods don't work with the `payload` object.
https://github.com/apache/cassandra-sidecar/blob/0a9b3f458d0788a1dd8d517d35bb4e4e4c1754a1/client-common/src/main/java/org/apache/cassandra/sidecar/common/request/data/CompactionStopRequestPayload.java#L77

Commit:
make minor structural changes, adjust tests

Commit:
throttle compaction via nodetool compactiontrhoughput to slow compactions for integration tests

Commit:
fix enum errors in integration tests

shalnisundram committed:
shorten integration test duration for triggering and testing compaction stop

shalnisundram committed:
eliminiate redundant tests and settings in CompactionStopIntegrationTest.java

jyothsnakonisa commented on adapters/adapters-base/src/test/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperationsTest.java:
```diff
@@ -0,0 +1,127 @@
+        verify(mockJmxClient, times(1))
+            .proxy(CompactionManagerJmxOperations.class, COMPACTION_MANAGER_OBJ_NAME);
+    }
+}
```
Please add a new line in the end of the file

jyothsnakonisa commented on client-common/src/test/java/org/apache/cassandra/sidecar/common/request/data/CompactionStopRequestPayloadTest.java:
```diff
@@ -0,0 +1,179 @@
+        assertThat(deser.compactionType()).isEqualTo(CompactionType.COMPACTION);
+        assertThat(deser.compactionId()).isEqualTo("Test-ID-123");
+    }
+}
```
Please add a new line at the end of the file

jyothsnakonisa commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,168 @@
+
+        return payload;
+    }
+}
```
Please add a new line here

jyothsnakonisa commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,168 @@
+                                                  CompactionStopRequestPayload request)
+    {
+        CompactionType compactionType = request.compactionType();
+        String compactionId = request.compactionId();
```
Do you want to trim the compactionID? I see that you are doing it in the other parts of the code

shalnisundram replied:
Won't trim here since in many cases the compactionId will be null, so want to avoid NPEs from attempting to trim in these scenarios. Have a null check and trim already in `hasValidCompactionId()` so will keep that structure
Though realized from this I wasn't returning the trimmed ID in `stopCompactionById()` so made that change there + respective tests

jyothsnakonisa commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,31 @@ public List<Map<String, String>> getCompactions()
+                    COMPACTION_MANAGER_OBJ_NAME
+            );
+            proxy.stopCompactionById(compactionId);
+        }
```
Do you want to throw an exception when the compactionId is invalid here and in the `stopCompaction` method

shalnisundram replied:
Done - thrown IllegalArgumentException. Thanks for catching that

jyothsnakonisa commented:
Looks good overall, few minor comments

shalnisundram committed:
return trimmed compaction ID, throw errors, add new lines

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,37 @@ public List<Map<String, String>> getCompactions()
+                    CompactionManagerJmxOperations.class,
+                    COMPACTION_MANAGER_OBJ_NAME
+            );
+            proxy.stopCompactionById(compactionId);
```
Should we use the trimmed `compactionId` here?

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -20,10 +20,12 @@
 import org.apache.cassandra.sidecar.adapters.base.jmx.CompactionManagerJmxOperations;
 import org.apache.cassandra.sidecar.common.server.CompactionManagerOperations;
 import org.apache.cassandra.sidecar.common.server.JmxClient;
+import org.jetbrains.annotations.NotNull;
```
Nit: Remove unused import

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,37 @@ public List<Map<String, String>> getCompactions()
+            );
+            proxy.stopCompactionById(compactionId);
+        } else {
+            throw new IllegalArgumentException("compaction id is null or empty");
```
Nit: can you also add the compaction id in error message. Easy to debug with that.

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,37 @@ public List<Map<String, String>> getCompactions()
+            proxy.stopCompaction(Objects.requireNonNull(compactionType,
+                    "compactionType must not be null when compactionId is not provided"));
+        } else {
+            throw new IllegalArgumentException("compaction type is null or empty");
```
Nit: can you also add the compaction type in error message. Easy to debug with that.

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/jmx/CompactionManagerJmxOperations.java:
```diff
@@ -34,4 +34,14 @@ public interface CompactionManagerJmxOperations
     List<Map<String, String>> getCompactions();
+
+    /**
+     * Stop compaction by type
```
Can you also that in Javadoc, that it throws exception for invalid compaction type, similarly javadoc for compaction by Id.

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/ApiEndpointsV1.java:
```diff
@@ -18,6 +18,8 @@
 
 package org.apache.cassandra.sidecar.common;
 
+import java.lang.ref.PhantomReference;
```
Nit: Remove unused import

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/request/data/CompactionStopRequestPayload.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    @JsonCreator
+    public CompactionStopRequestPayload(
+    @JsonProperty(value = "compaction_type") CompactionType compactionType,
```
Nit: for parameter names shall we use camel case, just for consistency?. Other request payloads use camel case

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStopResponse.java:
```diff
@@ -0,0 +1,170 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CompactionStopResponse
+{
+    public static final String COMPACTION_TYPE_KEY = "compaction_type";
```
Nit: here too shall we use camel case for parameter names for consistency?

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStopResponse.java:
```diff
@@ -0,0 +1,170 @@
+     * @param status         the status of the stop operation (e.g., "PENDING", "FAILED")
+     */
+    @JsonCreator
+    public CompactionStopResponse(@JsonProperty(COMPACTION_TYPE_KEY) CompactionType compactionType,
```
Nit: can you reuse these in request payload too. Feel free to ignore this comment.

sarankk commented on client-common/src/test/java/org/apache/cassandra/sidecar/common/request/data/CompactionStopRequestPayloadTest.java:
```diff
@@ -0,0 +1,179 @@
+import org.apache.cassandra.sidecar.common.data.CompactionType;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
```
Nit: Unused import.

sarankk commented on client/src/testFixtures/java/org/apache/cassandra/sidecar/client/SidecarClientTest.java:
```diff
@@ -2003,7 +2068,7 @@ void testNodeLifecycleInfo() throws Exception
         assertThat(result.currentState()).isEqualTo(CassandraState.RUNNING);
         assertThat(result.desiredState()).isEqualTo(CassandraState.RUNNING);
-        assertThat(result.status()).isEqualTo(OperationStatus.CONVERGED);
+        assertThat(result.status()).isEqualTo(Lifecycle.OperationStatus.CONVERGED);
```
Nit: These changes seem unrelated to PR.

sarankk commented on server-common/src/main/java/org/apache/cassandra/sidecar/common/server/CompactionManagerOperations.java:
```diff
@@ -32,4 +32,22 @@ public interface CompactionManagerOperations
+    void stopCompactionById(String compactionId);
+
+    /**
+     * Stops compaction based on compaction ID or type.
```
This comment is confusing. This method is only for stopping with compaction type right? similarly comment for `stopCompactionById` method.

shalnisundram replied:
Correct - clarification made in both Javadocs in most recent push

sarankk commented on server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java:
```diff
@@ -112,6 +115,28 @@ VertxRoute cassandraCompactionStatsRoute(RouteBuilder.Factory factory,
+            responseCode = "200",
+            content = @Content(mediaType = "application/json",
+                    schema = @Schema(implementation = CompactionStopResponse.class)))
+    @APIResponse(responseCode = "500", description = "Internal server error")
```
Ideally we should not see `500` response from handler. 500 means we did not foresee some error and handle it properly. I would suggest removing it from list of expected responses.

sarankk commented on server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java:
```diff
@@ -112,6 +115,28 @@ VertxRoute cassandraCompactionStatsRoute(RouteBuilder.Factory factory,
         return factory.buildRouteWithHandler(compactionStatsHandler);
     }
 
+    @POST  // Note: POST, not PUT, since we're sending data in the body
```
https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/PUT Http docs say we can send request content with `PUT` requests too. Difference between them is, `PUT` requests are idempotent. Gossip update route is an example https://github.com/apache/cassandra-sidecar/blob/6cf4b4c9c818efac825af999fb47ac545f75d430/server/src/main/java/org/apache/cassandra/sidecar/modules/CassandraOperationsModule.java#L395. Lets update this to `PUT` and remove note?

shalnisundram replied:
Yes would prefer an idempotent operation - thanks. Fix made

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CompactionType.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Supported compaction types based on Cassandra's OperationType enum
+ */
+public enum CompactionType
```
I do not see `MAJOR_COMPACTION` type here from `OperationType` class, is that intentional?

shalnisundram replied:
Thanks for the catch - I may have been looking at another C* version. Just changed it to reflect this list: https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/compaction/OperationType.java

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/data/CompactionType.java:
```diff
@@ -0,0 +1,74 @@
+    TOMBSTONE_COMPACTION,
+    ANTICOMPACTION,
+    VERIFY,
+    WRITE,
```
`WRITE` operation seems to be not a compaction type, can you double check?

sarankk reviewed (commented):
+1 Thanks Shalni, overall looks great to me. Left some comments, I am +1 once the comments are addressed.

shalnisundram committed:
change post to put request, clean up, edit compaction types

shalnisundram committed:
resolve checksytle errors

shalnisundram committed:
apply formatting changes

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/data/CompactionStopStatus.java:
```diff
@@ -0,0 +1,17 @@
+package org.apache.cassandra.sidecar.common.data;
```
Missing license for file

shalnisundram committed:
edit enums in tests, make bug fix for post to put request change

shalnisundram committed:
limit CompactionType functionality to adapter-specific checks for supported compaction type per cassandra version

limit CompactionType enum to use in cassandra version-specific adapters to check for supported compaction type

arjunashok commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,56 @@ public List<Map<String, String>> getCompactions()
+    {
+        return Arrays.stream(CompactionType.values())
+                .map(CompactionType::name)
+                .collect(Collectors.toList());
```
Optimize to return a `Set` for lookups, or optionally, a static set in the enum. 

eg.
```
    private static final List<String> NAMES = Arrays.stream(values())
            .map(CompactionType::name)
            .collect(Collectors.toUnmodifiableList());
```

arjunashok commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/CompactionType.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Supported compaction types based on Cassandra's OperationType enum
+ */
+public enum CompactionType
```
Let's be explicit about the version this corresponds to in the class name. eg. `Cassandra41CompactionType`

sarankk replied:
Supported compaction types seem to be same for 4.1 and 4.0, we can reuse 4.0's types instead of maintaining a separate list. Can you remove changes in 4.1 adpator classes?

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperations.java:
```diff
@@ -53,4 +55,56 @@ public List<Map<String, String>> getCompactions()
+    @Override
+    public List<String> supportedCompactionTypes()
+    {
+        return Arrays.stream(CompactionType.values())
```
Instead of creating a stream of values each time, can you compute it once and return it?

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraFactory.java:
```diff
@@ -31,9 +31,9 @@
+ * Factory to produce the Cassandra 5.x adapter
  */
-@MinimumVersion("4.0.0")
+@MinimumVersion("5.0.0")
```
`adapters-base` should still support 4.0.0 version, can you revert changes in this class? We will use the `Cassandra50Factory` you added for 5.0 version

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CompactionType.java:
```diff
@@ -0,0 +1,78 @@
+        }
+        catch (IllegalArgumentException unknownEnum)
+        {
+            String validTypes = String.join(", ",
```
Nit: create this list once

sarankk commented on adapters/adapters-base/src/test/java/org/apache/cassandra/sidecar/adapters/base/CassandraCompactionManagerOperationsTest.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    @Test
+    void testStopCompactionAllSupportedTypes()
```
Can you add a test case for unsupported compaction type?

sarankk commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/Cassandra41Adapter.java:
```diff
@@ -31,7 +32,7 @@
 
 /**
- * A {@link ICassandraAdapter} implementation for Cassandra 4.1 and later
+ * A {@link ICassandraAdapter} implementation for Cassandra 4.x
```
Supported compaction types list seem to be same between 4.0 and 4.1. We need not maintain `CompactionType` separately for 4.1, it can reuse 4.0's types. Can you revert changes in this class too?

sarankk commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/Cassandra41CompactionManagerOperations.java:
```diff
@@ -0,0 +1,56 @@
+ * Cassandra 4.1 implementation of CompactionManagerOperations that excludes MAJOR_COMPACTION
+ * which is not supported in Cassandra 4.x versions
+ */
+public class Cassandra41CompactionManagerOperations extends CassandraCompactionManagerOperations
```
Related to earlier comment, this class is not required.

sarankk commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/Cassandra41Factory.java:
```diff
@@ -31,9 +31,9 @@
+ * Factory to produce the Cassandra 4.x adapter
  */
-@MinimumVersion("4.1.0")
+@MinimumVersion("4.0.0")
```
We should keep the 4.1 factory as is. Can you revert changes here.

sarankk commented on adapters/adapters-cassandra50/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra50/Cassandra50Adapter.java:
```diff
@@ -0,0 +1,70 @@
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A {@link ICassandraAdapter} implementation for Cassandra 4.x
```
Can you update comment to say correct version?

sarankk commented on adapters/adapters-cassandra50/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra50/Cassandra50CompactionManagerOperations.java:
```diff
@@ -0,0 +1,56 @@
+     * Cassandra 5.x supports MAJOR_COMPACTION operation type
+     */
+    @Override
+    public List<String> supportedCompactionTypes()
```
👍

sarankk commented on adapters/adapters-cassandra50/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra50/Cassandra50StorageOperations.java:
```diff
@@ -0,0 +1,84 @@
+/**
+ * An implementation of the {@link StorageOperations} that interfaces with Cassandra 5.0 and later
+ */
+public class Cassandra50StorageOperations extends CassandraStorageOperations
```
can you extend from `Cassandra41StorageOperations`. You need not override the 2 methods if we extend from 4.1

sarankk commented on adapters/adapters-cassandra50/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra50/CompactionType.java:
```diff
@@ -0,0 +1,79 @@
+    ROW_CACHE_SAVE,
+    COUNTER_CACHE_SAVE,
+    INDEX_SUMMARY,
+    MAJOR_COMPACTION;
```
👍

sarankk commented on adapters/adapters-cassandra50/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra50/CompactionType.java:
```diff
@@ -0,0 +1,79 @@
+        }
+        catch (IllegalArgumentException unknownEnum)
+        {
+            String validTypes = String.join(", ",
```
Nit: compute this once.

sarankk commented on adapters/adapters-cassandra50/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra50/Cassandra50CompactionManagerOperations.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public List<String> supportedCompactionTypes()
+    {
+        return Arrays.stream(CompactionType.values())
```
Nit: compute this once

sarankk commented on server-common/src/main/java/org/apache/cassandra/sidecar/common/server/CompactionManagerOperations.java:
```diff
@@ -32,4 +32,23 @@ public interface CompactionManagerOperations
+     */
+    void stopCompaction(String compactionType);
+
+    List<String> supportedCompactionTypes();
```
Can you add java doc for this method too to add a note on how types can differ between versions

sarankk commented on client-common/src/main/java/org/apache/cassandra/sidecar/common/response/CompactionStopResponse.java:
```diff
@@ -0,0 +1,169 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CompactionStopResponse
+{
+    public static final String COMPACTION_TYPE_KEY = "compactionType";
```
Nit: can you use these String constants in request payload class too?

sarankk commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,175 @@
+        CompactionManagerOperations compactionManagerOps = metadataFetcher.delegate(host).compactionManagerOperations();
+
+        executorPools.service()
+                .executeBlocking(() -> stopCompaction(compactionManagerOps, request))
```
Nit: incorrect alignment

sarankk commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,175 @@
+        {
+            throw new IllegalArgumentException(
+                String.format("Compaction type '%s' not supported in this Cassandra version. " +
+                             "Supported types: %s",
```
Nit: incorrect alignment

sarankk commented on server/src/main/java/org/apache/cassandra/sidecar/handlers/CompactionStopHandler.java:
```diff
@@ -0,0 +1,175 @@
+        String compactionId = request.compactionId();
+
+        // Validate compaction type supported by this Cassandra version
+        if (compactionType != null && !operations.supportedCompactionTypes().contains(compactionType))
```
Since we are performing this check in stopCompaction call, we need not repeat it here again

sarankk reviewed (commented):
Changes look great to me, left final set of comments.

shalnisundram committed:
remove redundant CompactionType logic in cassandra41 adapter

sarankk commented on adapters/adapters-base/src/main/java/org/apache/cassandra/sidecar/adapters/base/CassandraFactory.java:
```diff
@@ -42,9 +41,9 @@ public class CassandraFactory implements ICassandraFactory
 
     public CassandraFactory(DnsResolver dnsResolver, DriverUtils driverUtils, TableSchemaFetcher tableSchemaFetcher)
     {
-        this.dnsResolver = Objects.requireNonNull(dnsResolver, "dnsResolver is required");
```
Can you revert these changes?

sarankk commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/Cassandra41Factory.java:
```diff
@@ -42,9 +41,9 @@ public class Cassandra41Factory implements ICassandraFactory
 
     public Cassandra41Factory(DnsResolver dnsResolver, DriverUtils driverUtils, TableSchemaFetcher tableSchemaFetcher)
     {
-        this.dnsResolver = Objects.requireNonNull(dnsResolver, "dnsResolver is required");
```
Can you revert these changes, they seem to be unrelated to PR

sarankk commented on adapters/adapters-cassandra41/src/main/java/org/apache/cassandra/sidecar/adapters/cassandra41/Cassandra41Adapter.java:
```diff
@@ -50,7 +50,7 @@ public Cassandra41Adapter(DnsResolver dnsResolver,
     @Override
     @NotNull
-    protected StorageOperations createStorageOperations(DnsResolver dnsResolver, JmxClient jmxClient)
+    public StorageOperations storageOperations()
```
Can you revert this change too

shalnisundram committed:
revert Cassandra41Adapter changes

shalnisundram committed:
add cassandra50 adapter to integration test framework

shalnisundram committed:
include cassandra5 build.gradle

shalnisundram committed:
fix checkstyle for integration tests

shalnisundram committed:
add cassandra50Adapter dependency to test context

shalnisundram committed:
add CASSIDECAR-360 compaction stop to CHANGES.txt

shalnisundram committed:
fix cassandra51Adapters CompactionType value stream bug

shalnisundram committed:
fix error handling for MAJOR_COMPACTION in integration tests

sarankk reviewed (approved):
+1 Thanks Shalni for great work.
